src/Core.Nhibernate/Stores/NhibernateStore.cs

```diff
@@ -24,71 +25,94 @@
 
 using System;
 using System.Data;
+using System.Threading.Tasks;
+using AutoMapper;
 using NHibernate;
 
 namespace IdentityServer3.Contrib.Nhibernate.Stores
 {
     public abstract class NhibernateStore
     {
+        protected readonly IMapper _mapper;
         private readonly ISession _nhSession;
 
-        protected NhibernateStore(ISession session)
+        protected NhibernateStore(ISession session, IMapper mapper)
         {
-            if (session == null) throw new ArgumentNullException(nameof(session));
-
-            _nhSession = session;
+            _mapper = mapper ?? throw new ArgumentNullException(nameof(mapper));
+            _nhSession = session ?? throw new ArgumentNullException(nameof(session));
         }
 
-        protected void ExecuteInTransaction(Action<ISession> actionToExecute, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
+        protected async Task<object> SaveAsync(object obj)
+            => await ExecuteInTransactionAsync(session => session.SaveAsync(obj));
+
+        protected async Task ExecuteInTransactionAsync(Func<ISession, Task> actionToExecute, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
         {
-            if (_nhSession.Transaction != null && _nhSession.Transaction.IsActive)
+            var transaction = _nhSession.GetCurrentTransaction();
+
+            if (transaction != null)
             {
-                actionToExecute.Invoke(_nhSession);
+                try
+                {
+                    await actionToExecute(_nhSession);
+                }
+                catch (Exception)
+                {
+                    await transaction.RollbackAsync();
+                    throw;
+                }
             }
             else
             {
                 using (var tx = _nhSession.BeginTransaction(isolationLevel))
                 {
                     try
                     {
-                        actionToExecute.Invoke(_nhSession);
-                        tx.Commit();
+                        await actionToExecute(_nhSession);
+                        await tx.CommitAsync();
                     }
                     catch (Exception)
                     {
-                        tx.Rollback();
+                        await tx.RollbackAsync();
                         throw;
                     }
                 }
 
             }
         }
 
-        protected T ExecuteInTransaction<T>(Func<ISession, T> actionToExecute, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
+        protected async Task<T> ExecuteInTransactionAsync<T>(Func<ISession, Task<T>> actionToExecute, IsolationLevel isolationLevel = IsolationLevel.ReadCommitted)
```

Comment:
When making these kinds of changes in a library you're better off adding the async version and marking the sync version `Obsolete`. In our case that would allow a 2-step update to Nearmap.IdentityServer to minimise risk.

Reply:
This method isn't directly used by our IdentityServer, for our use-case it's internal to this library only. I'll have a look at putting the non-asyncs back in though as an alternative for occasions where async is not desired for usage.